File: common-codegen/src/main/java/io/vrap/rmf/codegen/common/generator/util/PackageSwitch.java
package io.vrap.rmf.codegen.common.generator.util;

import io.vrap.rmf.raml.model.modules.Library;
import io.vrap.rmf.raml.model.resources.util.ResourcesSwitch;
import io.vrap.rmf.raml.model.types.Annotation;
import io.vrap.rmf.raml.model.types.AnnotationsFacet;
import io.vrap.rmf.raml.model.types.AnyType;
import io.vrap.rmf.raml.model.types.StringInstance;
import io.vrap.rmf.raml.model.types.util.TypesSwitch;
import org.eclipse.emf.ecore.EObject;
import org.eclipse.emf.ecore.util.ComposedSwitch;

import java.util.Optional;

public class PackageSwitch extends ComposedSwitch<String> {

    private final String basePackage;

    public PackageSwitch(final String basePackage) {
        this.basePackage = basePackage;
        addSwitch(new TypePackageSwitch());
        addSwitch(new ResourcePackageSwitch());
    }

    @Override
    public String defaultCase(EObject eObject) {
        return basePackage;
    }

    private class TypePackageSwitch extends TypesSwitch<String> {
        @Override
        public String defaultCase(EObject object) {
            return basePackage + ".models";
        }

        @Override
        public String caseAnyType(AnyType type) {
            final String modelsPackage = basePackage + ".models";
            while (type.getType() != null) {
                if ((type.eContainer() instanceof Library) && (((Library) type.eContainer()).getAnnotation("package") != null)) {
                    break;
                }
                type = type.getType();
            }
            EObject eContainer = type.eContainer();
            while (eContainer != null) {
                if (eContainer instanceof AnnotationsFacet) {
                    final String resultPackage = Optional.of(eContainer)
                            .map(AnnotationsFacet.class::cast)
                            .map(library -> library.getAnnotation("package"))
                            .map(Annotation::getValue)
                            .map(StringInstance.class::cast)
                            .map(StringInstance::getValue)
                            .map(s -> modelsPackage + "." + s)
                            .orElse(modelsPackage);
                    return resultPackage;
                }
                eContainer = eContainer.eContainer();
            }
            return modelsPackage;
        }

    }

    private class ResourcePackageSwitch extends ResourcesSwitch<String> {
        @Override
        public String defaultCase(EObject object) {
            return basePackage + ".client";
        }
    }



}

File: common-codegen-kt/src/main/kotlin/io/vrap/rmf/codegen/kt/types/PackageSwitch.kt
package io.vrap.rmf.codegen.kt.types

import com.google.inject.Inject
import com.google.inject.name.Named
import io.vrap.rmf.codegen.kt.di.VrapConstants
import io.vrap.rmf.raml.model.modules.Library
import io.vrap.rmf.raml.model.resources.util.ResourcesSwitch
import io.vrap.rmf.raml.model.types.AnnotationsFacet
import io.vrap.rmf.raml.model.types.AnyType
import io.vrap.rmf.raml.model.types.util.TypesSwitch
import org.eclipse.emf.ecore.EObject
import org.eclipse.emf.ecore.util.ComposedSwitch


class PackageSwitch @Inject constructor(@Named(VrapConstants.PACKAGE_NAME) val basePackage: String) : ComposedSwitch<String>() {

    init {
        addSwitch(TypePackageSwitch())
        addSwitch(ResourcePackageSwitch())
    }

    override fun defaultCase(eObject: EObject): String {
        return basePackage
    }

    private inner class TypePackageSwitch : TypesSwitch<String>() {
        override fun defaultCase(`object`: EObject?): String {
            return "$basePackage.models"
        }

        override fun caseAnyType(type: AnyType?): String {
            var currentType = type
            val modelsPackage = "$basePackage.models"
            while (currentType != null) {

                val annotation = currentType.getAnnotation("package") ?: currentType.type?.getAnnotation("package")
                if (annotation != null) {
                    return annotation.let { it.value }
                            ?.let { it.value }
                            ?.let { "$modelsPackage.$it" }
                            ?: modelsPackage
                }
                if (currentType.eContainer() is Library && (currentType.eContainer() as Library).getAnnotation("package") != null) {

                    var eContainer: EObject? = currentType.eContainer()
                    while (eContainer != null) {
                        if (eContainer is AnnotationsFacet) {
                            return eContainer.let { it.getAnnotation("package") }
                                    ?.let { it.value?.value }
                                    ?.let { "$modelsPackage.$it" }
                                    ?: modelsPackage
                        }
                        eContainer = eContainer.eContainer()

                    }
                }
                currentType = currentType.type
            }

            return modelsPackage
        }

    }

    private inner class ResourcePackageSwitch : ResourcesSwitch<String>() {
        override fun defaultCase(`object`: EObject): String {
            return "$basePackage.client"
        }
    }


}
